storage/subtree_cache.go

```diff
@@ -0,0 +1,168 @@
+package storage
+
+import (
+	"bytes"
+	"fmt"
+	"sync"
+
+	"github.com/google/trillian"
+)
+
+// GetSubtreeFunc describes a function which can return a Subtree from storage.
+type GetSubtreeFunc func(id NodeID) (*SubtreeProto, error)
+
+// SetSubtreeFunc describes a function which can store a Subtree into storage.
+type SetSubtreeFunc func(s *SubtreeProto) error
+
+// SubtreeCache provides a caching access to Subtree storage.
+type SubtreeCache struct {
+	// subtrees contains the Subtree data read from storage, and is updated by
+	// calls to SetNodeHash.
+	subtrees map[string]*SubtreeProto
+	// dirtyPrefixes keeps track of all Subtrees which need to be written back
+	// to storage.
+	dirtyPrefixes map[string]bool
+	mutex         *sync.RWMutex
+}
+
+// Suffix represents the tail of of a NodeID, indexing into the Subtree which
+// corresponds to the prefix of the NodeID.
+type Suffix struct {
+	// bits is the number of bits in the node ID suffix.
+	bits byte
+	// path is the suffix itself.
+	path []byte
+}
+
+func (s *Suffix) serialize() string {
+	r := make([]byte, 1, len(s.path)+1)
+	r[0] = s.bits
+	r = append(r, s.path...)
+	return string(r)
+}
+
+const (
+	// strataDepth is the depth of Subtree.
+	strataDepth = 8
+)
+
+// NewSubtreeCache returns a newly intialised cache ready for use.
+// TODO(al): consider supporting different sized subtrees - for now everything's subtrees of 8 levels.
+func NewSubtreeCache() SubtreeCache {
+	return SubtreeCache{
+		subtrees:      make(map[string]*SubtreeProto),
+		dirtyPrefixes: make(map[string]bool),
+		mutex:         new(sync.RWMutex),
+	}
+}
+
+// splitNodeID breaks a NodeID out into its prefix and suffix parts.
+// unless ID is 0 bits long, Suffix must always contain at least one bit.
+func splitNodeID(id NodeID) ([]byte, Suffix) {
+	if id.PrefixLenBits == 0 {
+		return []byte{}, Suffix{bits: 0, path: []byte{}}
+	}
+	prefixSplit := (id.PrefixLenBits - 1) / strataDepth
+	suffixEnd := (id.PrefixLenBits-1)/8 + 1
+	s := Suffix{
+		bits: byte((id.PrefixLenBits-1)%strataDepth) + 1,
+		path: make([]byte, suffixEnd-prefixSplit),
+	}
+	// TODO(al): is all this copying actually necessary?
+	copy(s.path, id.Path[prefixSplit:suffixEnd])
+	if id.PrefixLenBits%8 != 0 {
+		suffixMask := (byte(0x1<<uint((id.PrefixLenBits%8))) - 1) << uint(8-id.PrefixLenBits%8)
```

Comment:
I think we should be using look up tables for some of the bit flipping but don't worry about that now.